### PR TITLE
implement stroke style

### PR DESCRIPTION
**Pull request by dwelle**
fixes https://github.com/excalidraw/excalidraw/issues/1129

image

We may want to tweak the `dasharray` length and offsets, if we don't like the above.

For curves, such as lines, arrows, or circles, it tends to result in suboptimal lines:

image

Not sure what we can do about it. /cc @pshihn

### CODE REVIEW

**Comment by kbariotis** [comment]
Looks great I think 👍 For the suboptimal lines could we play around with the roughness property? maybe reduce it to 0 when stroke style is not solid? I think the double lines is what causing it.

And it would make more sense, especially in dashed or dotted lines to not have double lines. (We wouldn't make double dotted lines I think on paper, at least that's how I'm thinking of it).

image

**Comment by dwelle** [comment]
Good point. I've removed roughness for ellipses & multi-point lines/arrows, and reduced roughness for rectangles/diamonds (`0.5` looks good, but YMMV).

**Comment by dwelle** [comment]
Ideally, if roughjs supported to disable double-lines on case-by-case basis, that would be best.

**Comment by kbariotis** [comment]
Ah actually I noticed something @dwelle. The stroke style commands the inner filling as well. So the fillings are dashed/dotted as well. Can we maybe prevent that?

**Comment by dwelle** [comment]
> Ah actually I noticed something @dwelle. The stroke style commands the inner filling as well. So the fillings are dashed/dotted as well. Can we maybe prevent that?

Good catch. Mhm, that's a tough one since the fill and stroke is made in one go by roughjs.

**Comment by dwelle** [comment]
The only thing that I can think of is splitting the element into fill-only, and stroke-only, and doing a two-pass render. Not sure if it's not best fixed upstream in roughjs by @pshihn (or we can make a PR).

**Comment by kbariotis** [comment]
Yeah I noticed that the stroke width commands the inner filling as well. Maybe a separate issue to tackle it then since its an existing issue.

**Comment by dwelle** [comment]
On the upside, you can now do these fills (bold stroke width):

image

**Comment by pshihn** [comment]
I already had plans to allow ability to disable multiple strokes in roughjs.
I will add the ability to individually configure dashes as well. 
I'll add these features tonight hopefully :)

**Comment by dwelle** [comment]
> I already had plans to allow ability to disable multiple strokes in roughjs.

Super cool. :heart: 

> I will add the ability to individually configure dashes as well.
>
> Some non-excalidraw roughjs users have asked for an explicit dashed feature as well, so feels natural to fix it upstream

So maybe best to wait shipping this PR, even though we could fix it in a next one. Fine with me both ways.

If anyone wants to land this as it is, let's do this.

**Comment by kbariotis** [comment]
We can ofc wait a few days, no need to rush. Let’s pick it up tomorrow with clear heads 🙂

But I do think it would be safe to release even without that.

**Comment by lipis** [comment]
I think it shouldn't change the filling.. it looks strange.. just the stroke.

**Comment by pshihn** [comment]
@dwelle  roughjs 4.3.0  has these features now. New options when drawing a shape (https://github.com/pshihn/rough/wiki#strokelinedash):

** strokeLineDash** set line dash for stroking the shape

** fillLineDash** set line dash for for fill lines not the stroke

** disableMultiStroke** boolean property that will disable multi stroke effect 

** disableMultiStrokeFill** boolean property that will disable multi stroke effect  for filled lines

**Comment by dwelle** [comment]
Nice.. gonna try it out! :muscle:

**Comment by dwelle** [comment]
So, I used `strokeLineDash`, but had retain the manual workflow for arrows because we want to not use dashed strokes for arrow points.

I disabled multiStroke for ellipses and multi-point lines/arrows.

**Comment by pshihn** [comment]
Diamond doesnt seem to work.

**Comment by pshihn** [comment]
btw @dwelle  do the roughjs options seem to work properly as you expected? Any changes needed?

**Comment by dwelle** [comment]
Thanks for the catch, fixed. API works fine.

I'm thinking I'll just use `disableMultiStroke` for all non-solid strokes, and keep roughness. Previously I reduced roughness to reduce double stroke but kept it, for elements that didn't exhibit the problems like ellipses and arrows.

But I think keeping roughness and disabling multiStroke might look better.

**Comment by dwelle** [comment]
And increasing width by `0.5` for dashed lines makes it look similar to multiStroke:

image

**Comment by dwelle** [comment]
I'm satisfied. Ready to ship!

File: src/renderer/renderElement.ts [hunk 1]
```diff
@@ -229,9 +272,21 @@ function generateElement(
...
+          // we need to add dashed stroked for arrows manually outside rough,
+          //  because we want to prevent dashed stroke for arrow points
+          strokeLineDash:
+            element.type === "line" && element.strokeStyle === "dashed"
+              ? DASHARRAY_DASHED
+              : element.type === "line" && element.strokeStyle === "dotted"
+              ? DASHARRAY_DOTTED
+              : undefined,
+          roughness: element.roughness,
+          disableMultiStroke: element.strokeStyle !== "solid",
```
Comment (pshihn) [inline src/renderer/renderElement.ts:289]: Not sure when you need to manually do this here. When drawing the arrow head, do not set the line dash option in roughjs command.

Comment (pshihn) [reply src/renderer/renderElement.ts:289]: Later in this method, under the comment "add lines only in arrow", the options you pass to the `generator.line` function, unset the `strokeLineDash` option

Comment (dwelle) [reply src/renderer/renderElement.ts:289]: You're right. I wasn't aware we're drawing those manually.

**Comment by dwelle** [comment]
Make dots in dotted stroke shorter:

image

**Comment by dwelle** [comment]
Should we make arrow caps dotted, for dotted stroke?

image

**Comment by kbariotis** [comment]
Rough.js is up to date on master @dwelle.

> Should we make arrow caps dotted, for dotted stroke?

IMO, dotted caps look just fine. :)

**Comment by dwelle** [comment]
> Rough.js is up to date on master @dwelle

It wasn't until a while ago :).

> IMO, dotted caps look just fine. :)

Made them dotted, then.

**Comment by dwelle** [comment]
> Do we really want to change the fill as well.. it looks weird..

We're no longer doing that. Check the latest.

**Comment by dwelle** [comment]
On second review, there is something going on...

**Comment by dwelle** [comment]
Oh, I get it.. I increase stroke width by 0.5 for non-solid stroke elements, which also increases the fill stroke width :/.

**Comment by pshihn** [comment]
@dwelle  I think I know what the problem is. When filling, a shape roughjs needs `hachureGap` and `fillWeight` property (the gap between hachure lines and width of these lines). If you dont specify these, they are calculated from strokeWidth.

fillWeight = strokeWidth/2
hachureGap = strokeWidth * 4

**Comment by dwelle** [comment]
Done. Thanks for the pointers.

There's still some fill regeneration going on when switching between `solid` and `dashed`/`dotted`, not sure why. But now it at least keeps the pattern the same (in terms of width and gap), so it's not such a big deal.

**Comment by pshihn** [comment]
The regeneration is happening because you're changing the multistroke mode when going from solid to dashed -> Different number of lines in different order.

**Comment by dwelle** [comment]
> The regeneration is happening because you're changing the multistroke mode when going from solid to dashed -> Different number of lines in different order.

I see. I thought fill generation is independent of stroke generation. Or does multiStroke apply to fill, too? It doesn't look like it (visually).

**Comment by pshihn** [comment]
It doesn't but the seed is set per shape. So with the same seed you generate n random points vs generate 2n random points. I do not re-seed between fill and stroke. But something I could look into. This could also be tied to the other issue about double lines.  :/

**Comment by dwelle** [comment]
> This could also be tied to the other issue about double lines. :/

Came to my mind, too :).

**Comment by dwelle** [comment]
But, it's not really blocking this PR, so if everyone's ok, let's ship this. /cc @lipis

**Comment by kbariotis** [comment]
Shall we 🚢 this @dwelle ? It looks great, no blocking issues.

**Comment by dwelle** [comment]
In a little bit -- waiting for a suitable window for a tweet 😳

**Comment by jdestgermain** [comment]
So excited for this guys, great work!

**Comment by lipis** [comment]
Let's ship it! :)

**Pull request merged**